File: SRP Redux/Final/MarsRoverKata/Mars.cs
using System.Collections.Generic;
using System.Linq;

namespace MarsRoverKata
{
    public class Mars
    {
        public Size Bounds { get; private set; }
        public Point CenterOfThePlanet { get; private set; }

        private readonly List<Obstacle> _obstacles;
        public IReadOnlyList<Obstacle> Obstacles
        {
            get { return _obstacles; }
        }

        public Mars()
            : this(new Size(25, 25))
        { }

        public Mars(Size bounds)
        {
            Bounds = bounds;
            CenterOfThePlanet = new Point(Bounds.Width / 2, Bounds.Height / 2);
            _obstacles = new List<Obstacle>();
        }

        public void AddObstacle(Obstacle obstacle)
        {
            _obstacles.Add(obstacle);
        }

        public void RemoveObstacle(Obstacle obstacle)
        {
            _obstacles.Remove(obstacle);
        }

        public Point CalculateFinalPosition(Point from, Point desired)
        {
            Point newDestination = desired;
            newDestination = CalculatePositionY(desired, newDestination);
            newDestination = CalculatePositionX(desired, newDestination);

            if (!IsValidPosition(newDestination))
            {
                return from;
            }

            return newDestination;
        }

        private Point CalculatePositionX(Point desired, Point newDestination)
        {
            if (desired.X > Bounds.Width)
            {
                newDestination = new Point(0, desired.Y);
            }
            if (desired.X < 0)
            {
                newDestination = new Point(Bounds.Width, desired.Y);
            }
            return newDestination;
        }

        private Point CalculatePositionY(Point desired, Point newDestination)
        {
            if (desired.Y > Bounds.Height)
            {
                newDestination = new Point(desired.X, 0);
            }
            if (desired.Y < 0)
            {
                newDestination = new Point(desired.X, Bounds.Height);
            }
            return newDestination;
        }

        public bool IsValidPosition(Point point)
        {
            bool anyInstance = _obstacles.Any(x => x.Location.Equals(point));

            return !anyInstance;
        }
    }
}

File: LSP+ISP/Java_Exercise/MarsRoverKata/src/main/java/MarsRoverKata/Mars.java
package MarsRoverKata;

import com.google.common.base.Predicate;
import com.google.common.collect.Collections2;
import com.google.common.collect.Iterables;

import java.util.ArrayList;
import java.util.List;

public class Mars {
    private AlienBehavior alienBehavior; // get/set
    private Size bounds; // get
    private Point centerOfThePlanet; // get

    private final List<IObstacle> obstacles;

    public List<IObstacle> getObstacles() {
        return obstacles;
    }

    public Mars() {
        this(new Size(25, 25));
    }

    public Mars(Size bounds) {
        this.bounds = bounds;
        alienBehavior = new AlienBehavior();
        centerOfThePlanet = new Point(this.bounds.getWidth() / 2, this.bounds.getHeight() / 2);
        obstacles = new ArrayList<IObstacle>();
    }

    public AlienBehavior getAlienBehavior() {
        return alienBehavior;
    }

    public void setAlienBehavior(AlienBehavior alienBehavior) {
        this.alienBehavior = alienBehavior;
    }

    public Size getBounds() {
        return bounds;
    }

    public Point getCenterOfThePlanet() {
        return centerOfThePlanet;
    }

    public void updateAliens() {
        for (IObstacle alien : Collections2.filter(getObstacles(), new Predicate<IObstacle>() {
            @Override
            public boolean apply(IObstacle input) {
                return input instanceof Alien;
            }
        })) {
            alienBehavior.moveAlien((Alien) alien);
        }
    }

    public void addObstacle(IObstacle obstacle) {
        obstacles.add(obstacle);
    }

    public void removeObstacle(IObstacle obstacle) {
        obstacles.remove(obstacle);
    }

    public Point calculateFinalPosition(Point from, Point desired) {
        Point newDestination = desired;
        newDestination = calculatePositionY(desired, newDestination);
        newDestination = calculatePositionX(desired, newDestination);

        if (!isValidPosition(newDestination)) {
            return from;
        }

        return newDestination;
    }

    private Point calculatePositionX(Point desired, Point newDestination) {
        if (desired.getX() > bounds.getWidth()) {
            newDestination = new Point(0, desired.getY());
        }
        if (desired.getX() < 0) {
            newDestination = new Point(bounds.getWidth(), desired.getY());
        }
        return newDestination;
    }

    private Point calculatePositionY(Point desired, Point newDestination) {
        if (desired.getY() > bounds.getHeight()) {
            newDestination = new Point(desired.getX(), 0);
        }
        if (desired.getY() < 0) {
            newDestination = new Point(desired.getX(), bounds.getHeight());
        }
        return newDestination;
    }

    public boolean isValidPosition(final Point point) {
        boolean anyInstance = Iterables.any(obstacles, new Predicate<IObstacle>() {
            @Override
            public boolean apply(IObstacle input) {
                return input.getLocation().equals(point);
            }
        });
        return !anyInstance;
    }
}

